src/logics/zeroex-v4/logic.swap-token.ts

```diff
@@ -0,0 +1,141 @@
+import * as common from '@protocolink/common';
+import { axios } from 'src/utils/http';
+import { getTokenList as getTokenListBase } from 'src/utils';
+import invariant from 'tiny-invariant';
+import { isWrapOrUnwrap } from 'src/logics/zeroex-v4/utils';
+import { slippageToZeroEx } from 'src/logics/zeroex-v4/slippage';
+
+export type SwapTokenLogicParams = core.TokenToTokenExactInParams<{
+  apiKey: string;
+  slippage?: number;
+  excludedSources?: string[];
+  includedSources?: string[];
+  skipValidation?: boolean;
+  takerAddress?: string;
+}>;
+
+export type SwapTokenLogicFields = core.TokenToTokenExactInFields<{
+  apiKey: string;
+  slippage?: number;
+  excludedSources?: string[];
+  includedSources?: string[];
+  skipValidation?: boolean;
+  takerAddress?: string;
+}>;
+
+export type ZeroExQuote = {
+  buyAmount: string;
+  data: string;
+  to: string;
+  expectedSlippage?: string;
+};
+
+export type SwapTokenLogicOptions = Pick<core.GlobalOptions, 'account'>;
+
+export class SwapTokenLogic
+  extends core.Logic
+  implements core.LogicTokenListInterface, core.LogicOracleInterface, core.LogicBuilderInterface
+{
+  static id = 'swap-token';
+  static protocolId = 'zeroex-v4';
+  static readonly supportedChainIds = supportedChainIds;
+
+  async getTokenList() {
+    return getTokenListBase(getTokenListUrls(this.chainId), this.chainId, [this.nativeToken]);
+  }
+
+  getAPIBaseUrl(chainId: number) {
+    switch (chainId) {
+      case common.ChainId.mainnet:
+        return 'https://api.0x.org/';
+      case common.ChainId.optimism:
+        return 'https://optimism.api.0x.org/';
+      case common.ChainId.polygon:
+        return 'https://polygon.api.0x.org/';
+      case common.ChainId.base:
+        return 'https://base.api.0x.org/';
+      case common.ChainId.arbitrum:
+        return 'https://arbitrum.api.0x.org/';
+      case common.ChainId.avalanche:
+        return 'https://avalanche.api.0x.org/';
+      default:
+        throw new Error('Unsupported chain');
+    }
+  }
+
+  getAPIHeaders(apiKey: string) {
+    return {
+      '0x-api-key': apiKey,
+    };
+  }
+
+  async quote(params: SwapTokenLogicParams) {
+    try {
+      const { input, tokenOut, slippage, excludedSources, includedSources, apiKey, takerAddress, skipValidation } =
+        params;
+      const slippagePercentage = slippage != null ? slippageToZeroEx(slippage) : undefined;
+      const url = this.getAPIBaseUrl(this.chainId) + `swap/v1/price`;
+      const {
+        data: { buyAmount },
+      } = await axios.get<ZeroExQuote>(url, {
+        params: {
+          slippagePercentage,
+          sellToken: input.token.elasticAddress,
+          buyToken: tokenOut.elasticAddress,
+          sellAmount: input.amountWei.toString(),
+          excludedSources: excludedSources?.join(','),
+          includedSources: includedSources?.join(','),
+          takerAddress,
+          skipValidation,
```

Comment:
Since the `takerAddress` acts as the user's agent which is smart contract, I think we can allow both the `quote` and `build` functions to calculate the agent address and assign it to `takerAddress`. Ref: https://github.com/dinngo/protocolink-logics/blob/master/src/logics/openocean-v2/logic.swap-token.ts#L77

Additionally, seems that `skipValidation` can be set to `true` for the `quote` endpoint, given that the user's agent is smart contract. Ref: 
- https://0x.org/docs/0x-swap-api/guides/accessing-rfq-liquidity/how-to-integrate-rfq-liquidity#quote-validation.
- https://0x.org/docs/0x-swap-api/api-references/get-swap-v1-price#request
  - For `/price`, the default of `skipValidation=true`
  - for `/quote`, the default of `skipValidation=false`

Reply:
Yeah, that's a good idea actually, will do that later today.